src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultImmutableConverter.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Reflection;
+using System.Text.Json.Serialization.Policies;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class DefaultImmutableConverter : JsonEnumerableConverter
+    {
+        public const string ImmutableNamespace = "System.Collections.Immutable";
+
+        private const string ImmutableListGenericTypeName = "System.Collections.Immutable.ImmutableList`1";
+        private const string ImmutableListGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableList`1";
+
+        private const string ImmutableArrayGenericTypeName = "System.Collections.Immutable.ImmutableArray`1";
+
+        private const string ImmutableStackGenericTypeName = "System.Collections.Immutable.ImmutableStack`1";
+        private const string ImmutableStackGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableStack`1";
+
+        private const string ImmutableQueueGenericTypeName = "System.Collections.Immutable.ImmutableQueue`1";
+        private const string ImmutableQueueGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableQueue`1";
+
+        private const string ImmutableSortedSetGenericTypeName = "System.Collections.Immutable.ImmutableSortedSet`1";
+
+        private const string ImmutableHashSetGenericTypeName = "System.Collections.Immutable.ImmutableHashSet`1";
+        private const string ImmutableSetGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableSet`1";
+
+        // This method is invoked with reflection in CreateFromList.
+        private static IEnumerable CreateFromListInternal<T>(MethodInfo createImmutableMethod, IList sourceList)
+        {
+            List<T> list = new List<T>();
+
+            foreach (object item in sourceList)
+            {
+                if (item is T itemT)
+                {
+                    list.Add(itemT);
+                }
+            }
+
+            return (IEnumerable)createImmutableMethod.Invoke(null, new object[] { list } );
+        }
+
+        public override IEnumerable CreateFromList(Type enumerableType, Type elementType, IList sourceList)
+        {
+            Debug.Assert(enumerableType.IsGenericType);
+
+            Type underlyingType = enumerableType.GetGenericTypeDefinition();
```

Comment:
**ALL** of this kind of reflection should happen once and be cached on a contract that represents what you want to serialize

e.g. https://github.com/JamesNK/Newtonsoft.Json/blob/4ab34b0461fb595805d092a46a58f35f66c84d6a/Src/Newtonsoft.Json/Serialization/JsonArrayContract.cs

Reflection like this is a cost that should be paid once. It doesn't change.

Reflection that happens per-serialization, e.g. the calling constructor, should use IL emit to generate high performance methods that are cached and called each serialization.

Reply:
Right now I've cached delegates for these reflection-constructed methods, and added a todo to use IL emit.